doc/protocol.md

```diff
@@ -0,0 +1,259 @@
+  current data stream.  If a request is resumed, it should not be assume that
+  the `typeKey` values will be the same as the original stream.  The
+  `typeName` parameter is a full qualified record name `package.message` that
+  identifies the type.  All GA4GH objects will match `org.ga4gh.*`.  Non-GA4GH
+  data types may also use this protocol, however they must not be under the
+  `org.ga4gh` package.
+
+- `dataObject typeKey data` - Used to send a data object.  The `typeKey`
+  parameter is used to find the matching `dataType` message.  The `data`
+  parameter is the data object.
+
+- `checkpoint memento` - Specify a checkpoint in the data stream.
+  The `memento` is an opaque, server-dependent object.  It should be
+  returned as-is when resuming a stream.
+
+- `resume memento` - Resume a transfer at the specified checkpoint, as define
+  by the `memento` object, in data stream.  The next data object that was or
+  would have been received after  the corresponding `checkpoint` message
+  will be the next data object received on the stream. It will be preceded by
+  the required `dataType` message.  The same checkpoint may resumed multiple
+  times, as would be required if there was a failure before another
+  `checkpoint` messages is received in the stream.
+
+## Message Encoding
+Messages are encoded using
+[Protocol Buffers V3](https://developers.google.com/protocol-buffers/docs/proto3).
+Use of protocol buffers to encode a message is independent of the encoding
+of the GA4GH object.  That is, there are two levels of Protocol Buffers
+encoding, the object itself, which are send in GA4GH messages, also encoded
+using Protocol Buffers.
+
+Messages are encoded as `oneof` alternatives in a `Message` object.  The Protocol
+Buffers declarations for the GA4GA messages are:
+
+```
+    syntax = "proto3";
+
+    package org.ga4gh.protocol;
+
+    message DataType {
+      int32 type_key = 1;
+      string type_name = 2;
+    }
+
+    message DataObject {
+      int32 type_key = 1;
+      bytes data = 2;
+    }
+
+    message Checkpoint {
+      bytes memento = 1;
+    }
+
+    message Restart {
+      bytes memento = 1;
+    }
+
+    message Message {
+      oneof message_type {
+        DataType data_type = 1;
+        DataObject data_object = 2;
+        Checkpoint checkpoint = 3;
+        Restart restart = 4;
+      }
+    }
+```
+
+
+### JSON ASCII message encoding
+
+Messages are encoding as UTF-8 text following the standard defined by
+Protocol Buffers V3 JSON encoding.  While implementation
+are not required to use the Protocol Buffers software, the JSON
+encode must follow its specification.   Messages are written to
+the stream in [line-delimited JSON format]
+(https://en.wikipedia.org/wiki/JSON_Streaming#Line_delimited_JSON).
+
+The MIME type for JSON encode is `application/ga4gh.v${apiversion}+JSON`.
+
+### Binary message encoding
+Message may be encoded in an efficient binary format using Protocol Buffers V3
```

Comment:
Doesn't this conflict with the Message Encoding section above, where we say the encapsulation is protobuf? Or am I misunderstanding this?